### PR TITLE
Make key value storage abstractions more accurate

### PR DESCRIPTION
**Pull request by adaszko**
Continuation of the chain of database abstractions refactorings.  Implements the idea at https://github.com/sigp/lighthouse/pull/1200#issuecomment-635192106.

Most of the tests were based on `MemoryStore` instead of `HotColdDB` due to efficiency.  This has an undesirable side effect of using different code to access the database in tests and in production, reducing test coverage and potentially leaving some logic bugs specific to database 'layering' (as done by `HotColdDB`) undetected in tests.  This PR proposes a solution that brings together best of both worlds: `HotColdDB` in tests is based on two layers of `MemoryStore`s so that it doesn't have to touch the disk, while at the same time still exercises the layering logic.  This is achieved via parameterization of `HotColdDB` with "hot" and "cold" kv stores types that get plugged with desired types depending on where `HotColdDB` gets used: `HotColdDB<MemoryStore, MemoryStore>` in tests, and `HotColdDB<LevelDB, LevelDB>` in production.  Thanks to that change the `Store` trait, which over time grew incoherent, no longer serves any purpose, so it gets removed as well.

### CODE REVIEW

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 1]
```diff
@@ -75,24 +74,22 @@ impl StoreItem for SszEth1 {
...
+    T: Eth1ChainBackend<E>,
     E: EthSpec,
-    S: Store<E>,
 {
     backend: T,
     /// When `true`, the backend will be ignored and dummy data from the 2019 Canada interop method
     /// will be used instead.
     pub use_dummy_backend: bool,
-    _phantom: PhantomData<(E, S)>,
+    _phantom: PhantomData<E>,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/eth1_chain.rs:86]: Great simplification!

File: beacon_node/beacon_chain/src/test_utils.rs [hunk 2]
```diff
@@ -192,7 +195,11 @@ impl<E: EthSpec> BeaconChainHarness<DiskHarnessType<E>> {
             .logger(log.clone())
             .custom_spec(spec)
             .store(store.clone())
-            .store_migrator(<BlockingMigrator<_> as Migrate<E>>::new(store, log.clone()))
+            .store_migrator(<BlockingMigrator<_, LevelDB<E>, LevelDB<E>> as Migrate<
+                E,
+                LevelDB<E>,
+                LevelDB<E>,
+            >>::new(store, log.clone()))
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/test_utils.rs:202]: Are all these annotations necessary? I think some could be underscores

File: beacon_node/beacon_chain/src/eth1_chain.rs (hunk 1 above)
Comment (adaszko) [reply beacon_node/beacon_chain/src/eth1_chain.rs:86]: Thanks :)

**Comment by michaelsproul** [comment]
Looks like there's a CI issue related to deps, you might need to add back in the `sloggers` dep to `beacon_chain` (it was possibly removed as part of the dep cleanup)

**Pull request merged**